packages/common/src/utils/urlFormatting.ts

```diff
@@ -25,6 +26,48 @@ export const createURL = (url: string, queryParams: QueryParamsArg) => {
   return `${baseUrl}${queryString ? `?${queryString}` : ''}`;
 };
 
+type ExtractRouteParams<T extends string> = T extends `${infer _Start}:${infer Param}/${infer Rest}`
+  ? { [K in Param | keyof ExtractRouteParams<Rest>]: string }
+  : T extends `${infer _Start}:${infer Param}`
+  ? { [K in Param]: string }
+  : object;
+
+type PathParams<T extends string> = T extends `${infer _Start}*` ? ExtractRouteParams<T> & Record<string, string | undefined> : ExtractRouteParams<T>;
+
+// Creates a route path from a path string and params object
+export const createPath = <Path extends string>(originalPath: Path, pathParams?: PathParams<Path>, queryParams?: QueryParamsArg): string => {
+  const path = originalPath
+    .split('/')
+    .map((segment) => {
+      if (segment === '*') {
+        // Wild card for optional segments
+        if (!pathParams) return segment;
+
+        return Object.entries(pathParams)
+          .filter(([key]) => !originalPath.includes(key))
+          .map(([_, value]) => value)
+          .join('/');
+      }
+      if (!segment.startsWith(':') || !pathParams) return segment;
+
+      const isOptional = segment.endsWith('?');
+      const paramName = segment.replace(':', '').replace('?', '');
+      const paramValue = pathParams[paramName as keyof typeof pathParams];
```

Comment:
There are quite some tricks happening here and I cannot understand why these exceptional cases are needed. Maybe they needs explaining in the code? Or can you explain in a comment?

Eg. param names with a : or ? seems a bit weird for me.

Reply:
This is about route path definitions, they contain these characters to make the optional or required. See `paths.tsx`.
I do understand that it is a bit complex. Any suggestions how to make things clear? Which aspect raises most questions?

Reply:
I think you cannot really simplify this more because the path definitions can indeed contain these `:` or `?` characters. But I do agree with Mike on the fact that it is not clear why this is necessary, so that could be explained in a code comment

Reply:
Adding some code comments is sufficient enough for me

Reply:
Nice!